File: wordle/solver.py
#!/usr/bin/env python3
# 
#   solver.py 
#   Samir Gupta
#   05/03/2022
#
from guesses import guesses
from answers import answers
from time import time
from random import shuffle
both = guesses + answers

class Game:
    def __init__(self, first_guess, bot=False, n=1):
        # initialising with bot=True, makes the computer solve n words
        # initialising with bot=False, is user mode, where the user inputs the outcomes
        self.guess = first_guess
        self.bot = bot
        self.answers = answers
        self.n = 1
        self.outcomes = total_outcomes()
        if self.bot:
            self.time = time()
            self.rand_ans = [a for a in answers]
            shuffle(self.rand_ans)
            self.rand_n = 0
            self.errors = 0
            self.attempts = 0
            self.words = 0
            self.total = n
            self.rand_word()


    def rand_word(self):
        # getting a random word in answers
        self.word = self.rand_ans[self.rand_n]
        self.rand_n += 1
        if self.rand_n == len(self.rand_ans):
            self.rand_n = 0
            shuffle(self.rand_ans)
        print(f'\n{self.word.upper()} - {self.words+1}/{self.total}')
    

    def check(self, guess):
            # checking a guess against the word, returns the outcome of the guess

            # default outcome
            outcome = ['-', '-', '-', '-', '-']

            # keeping track of the letters that have been checked (for doubles)
            letters = {}

            # get g's and add to letters
            for x in range(len(guess)):
                if guess[x] == self.word[x]:
                    outcome[x] = 'g'
                    try: 
                        letters[guess[x]] += 1
                    except:
                        letters[guess[x]] = 1
            
            # getting o's if there are any while taking into consideration doubles
            for x in range(len(guess)):
                count = self.word.count(guess[x])
                if count and outcome[x] != 'g':
                    try:
                        if letters[guess[x]] < count:
                            outcome[x] = 'o'
                            letters[guess[x]] += 1
                    except:
                        outcome[x] = 'o'
                        letters[guess[x]] = 1
            return outcome


    def solve_word(self):
        # function used to solve for a word

        # guess is printed (extra space is to make sure it clears the line when printed)
        print(f'GUESS : {self.guess.upper()}              ')
        self.n += 1

        # result of guess is obtained
        if self.bot:
            self.res = self.check(self.guess)
            print('RESULT : ' + ''.join(self.res))
        else:
            self.res = list(input(f"RESULT : "))
            while not outcome_check(self.res):
                print(f'GUESS : {self.guess.upper()}              ')
                self.res = list(input(f"RESULT : "))

        # answers are refined to meet result
        self.new_ans = list(filter(lambda an: is_possible(an, self.guess, self.res), self.answers))

        # if guess was correct (without refining answers down to one) reduce the amount of guesses and dont print answers left
        if self.res == ['g', 'g', 'g', 'g', 'g']:
                self.n -= 1
        else:
            print(f'ANSWERS LEFT : {len(self.new_ans)}')
        
        # if one answer is left, game is won, if none are left, there was an error
        if len(self.new_ans) in (0, 1):
            self.word_complete(bool(len(self.new_ans)))
            return 
    
        # using next_guess.txt to find the second guess after trace
        if self.guess == 'trace':
            with open('next_guess.txt', 'r') as f:

                # go through each line in the file
                for line in f.readlines():

                    # get the outcome from the line
                    split = line[:-1].split(' - ')
                    o_check = [split[0][i] for i in range(len(split[0])) if split[0][i] in ('g', 'o', '-')]

                    # if the outcome from the line is the one got when 'trace' was used, bypass solving for the second guess and go straight to the third
                    if o_check == self.res:
                        self.answers = self.new_ans
                        self.guess = split[1]
                        self.solve_word()
                        return 
        
        # hold a temporary best
        best = ('', len(self.new_ans))

        # check each guess
        for g in range(len(both)):
            total = len(self.outcomes)
            possible_ans = 0
            for o in self.outcomes: # check each outcome

                # refine the answers to get the amount of words left
                ans_left = len(list(filter(lambda a: is_possible(a, both[g], o), self.new_ans)))
                if not ans_left:
                    total -= 1
                possible_ans += ans_left
            
            # the temporary best is compared to the current guess
            best = min(best, (both[g], possible_ans/total), key=lambda x: x[1])

            print(f'Checking {g+1}/{len(both)}', end='\r')

        # next guess is determined and solving continues
        self.answers = self.new_ans
        self.guess = best[0]
        self.solve_word()


    def word_complete(self, success):
        # to be run when a word is either solved or an error occured

        # print output to user and update variables
        if success:
            print(f'** WORD = {self.new_ans[0].upper()} - {self.n} Guesses **')
            if self.bot:
                self.attempts += self.n
            else:
                return
        elif not success and self.bot:
            print(f'ERROR OCCURED')
            self.errors += 1
        else:
            # if error occurs, print all possible answers before error occured (user mode)
            print(f'Error Occured, answer is one of : {self.answers}')
            return

        # end word guessing if total is reached, print summary (bot only)
        self.words += 1
        if self.words == self.total:
            print(f'\nCompleted {self.total} Words\nAverage Guesses: {round(self.attempts/(self.total-self.errors), 2)}\nErrors: {self.errors}\nTime : {round((time()-self.time)/60, 2)} mins')
            return
        
        # reinitialise variables, get a new word and solve it (bot only)
        self.n = 1
        self.answers = answers
        self.guess = 'trace'
        self.rand_word()
        self.solve_word()


def outcome_check(outcome):
    # determine if outcome is viable
    for char in outcome:
        if char not in ('-', 'o', 'g'):
            print("ERROR : input can only contain '-', 'o' or 'g'")
            return False

    if len(outcome) != 5:
        print("ERROR : input has to be 5 characters long")
        return False
    
    return True


def total_outcomes(n=5, part=[]):
    # function to get all possible outcomes
    if len(part) == n:
        return [part]
    else:
        res = []
        for o in ['-', 'o', 'g']:
            res += total_outcomes(n, part + [o])
        return res


def is_possible(word, guess, outcome):
    # determines whether a word is viable to be an answer, given an attempt and the outcome the attempt provided

    # split the indexes into groups [[g's], [o's], [-'s]] as checking requires g's to be checked first, o's second and -'s last
    order = [[],[],[]]
    for i in range(len(outcome)):
        if outcome[i] == 'g':
            order[0].append(i)
        elif outcome[i] == 'o':
            order[1].append(i)
        else:
            order[2].append(i)

    # keeping tracking of letters that have been checked (for doubles)
    checked = {}
    for group in order:
        for x in group: 
            # go through the g's in outcome first, then the o's, then the -'s

            count = word.count(guess[x])

            # if the outcome is g, check if word contains guess[x] at position x
            # add to checked
            if outcome[x] == 'g':
                if word[x] != guess[x]:
                    return False
                try:
                    checked[guess[x]] += 1
                except:
                    checked[guess[x]] = 1

            # if outcome is o, if checked[guess[x]] exists, it must be less than count and guess[x] cannot equal word[x] in order for the word to be viable
            # if checked[guess[x]] doesnt exist, there must be a count and guess[x] cannot equal word[x] for the word to be viable
            elif outcome[x] == 'o':
                try:
                    if checked[guess[x]] == count or guess[x] == word[x]:
                        return False
                    checked[guess[x]] += 1
                except:
                    if not count or guess[x] == word[x]:
                        return False
                    checked[guess[x]] = 1

            # if the outcome is -, if checked[guess[x]] exists and there is a count, checked[guess[x]] must already be equal to count for the word to be viable
            # if checked[guess[x]] doesnt exist but there is a count the word cannot be viable
            elif outcome[x] == '-':
                if count:
                    try:
                        if checked[guess[x]] < count:
                            return False
                    except:
                        return False
    return True


def get_first_guess():
    # code used to produce first_guess.txt and find that 'trace' was the best first word
    outcomes = total_outcomes()
    word_set = []
    for g in range(len(both)):
        total = len(outcomes)
        possible_ans = 0
        for o in outcomes: # check all outcomes

            # see how many answers are left are both[g] is guessed
            ans_left = len(list(filter(lambda a: is_possible(a, both[g], o), answers)))
            
            # if 0 answers are left, outcome is impossible therefore excluded from total
            if not ans_left:
                total -= 1
            possible_ans += ans_left

        # get average, print for reference and append to word_set
        av = possible_ans/total
        print(f'{g+1} - {both[g]} - {av}') 
        word_set.append((both[g], av))

    # create file to write the sorted word_set in
    with open('first_guess.txt', 'w') as f:
        for x in sorted(word_set, key=lambda w: w[1]):
            f.write(f'{x[0]} - {x[1]}\n')


def get_next_word():
    # finding the next guess after 'trace' by working out the best possible word depending on the 243 outcomes
    first_guess = 'trace'
    outcomes = total_outcomes()
    for o1 in reversed(outcomes):
        print(o1)
        ans_mid = list(filter(lambda a: is_possible(a, first_guess, o1), answers))
        best = ('', len(ans_mid))
        for w in range(len(both)):
            total = len(outcomes)
            possible_ans = 0
            for o2 in outcomes: # check all outcomes
                ans_left = len(list(filter(lambda a: is_possible(a, both[w], o2), ans_mid)))
            
                if not ans_left:
                    total -= 1
                possible_ans += ans_left

            if total != 0:
                best = min(best, (both[w], possible_ans/total), key=lambda x: x[1])

            print(f'{w+1}/{len(both)}', end='\r')

        # print to text file
        with open('next_guess.txt', 'a') as f:
            f.write(f'{o1} - {best[0]}\n')


if __name__ == '__main__':
    MODE = input('MODE: ').upper()
    while MODE not in ('USER', 'BOT'):
        print('MODE INVALID : Must be either USER or BOT')
        MODE = input('MODE: ').upper()
    
    if MODE == 'BOT':
        n = input("WORDS: ")        
        while 1:
            try:
                n = int(n)
                if n > 0 and n <= 200: # 200 is set as a limit to prevent exceeding recursion depth
                    break
            except:
                pass
            
            print('WORDS INVALID : Must be an integer n where 1 <= n <= 200')
            n = input("WORDS: ")

        wordle = Game('trace', True, n)
        wordle.solve_word()
    else:
        wordle = Game('trace')
        wordle.solve_word()

File: wordle/solver.c
/**
 * Solver for Wordle
 * @author Samir Gupta
 * @date 2022-09-11
 * @note This implementation was originally translated from Python
 */

#include <stdio.h>
#include <stdbool.h>
#include <string.h>

#define ANSWERS_FILE "answers.txt" // answers file name
#define GUESSES_FILE "guesses.txt" // guesses file name
#define ANSWER_LENGTH 2315         // the amount of valid answers wordle has
#define GUESSES_LENGTH 12972       // the amount of valid guesses wordle allows

#define WORD_SIZE 6      // the total word size of each word (5 + escaping)
#define ALPHABET_SIZE 26 // the total size of the alphabet the words derive from

#define TOTAL_OUTCOMES 243 // the total number of outcomes possible in wordle
#define OUTCOMES_LENGTH 3  // the total number of options for each letter in an outcome
#define GREEN_INPUT 'g'    // the green input character
#define ORANGE_INPUT 'o'   // the orange input character
#define BLACK_INPUT '-'    // the black input character
#define GREEN 2            // the green integer used in replacement of the green character
#define ORANGE 1           // the orange integer used in replacement of the orange character
#define BLACK 0            // the black integer used in replacement of the black character

#define NULL_INDEX -1 // index used to identify when we've hit the end of an integer array

/**
 * Counts the number of a particular character in a string.
 * @param string The string to look in.
 * @param character The character to look for.
 * @return The number of this character in the string.
 */
int count_chr(char *string, char character)
{
    int count = 0;
    for (int i = 0; i < WORD_SIZE; i++)
    {
        if (string[i] == character)
            count++;
    }
    return count;
}

/**
 * Converts a character into it's lowercase equivalent.
 * @param character The character to convert.
 * @return The converted character.
 */
char to_lower(char character)
{
    if (character >= 'A' && character <= 'Z')
    {
        return character + 'a' - 'A';
    }
    return character;
}

/**
 * Get the first guess from the user's input.
 * @param guesses The valid guesses the user can choose from.
 * @return The user's first guess.
 */
char *get_first_guess(char guesses[][WORD_SIZE])
{
    static char first_guess[WORD_SIZE];
    printf("\n");
    while (1)
    {
        printf(" GUESS: ");

        for (int i = 0; i < WORD_SIZE - 1; i++) // get five characters
        {
            char c = to_lower(getchar());
            if (c == '\n') // skip end line character
            {
                i--;
                continue;
            }
            first_guess[i] = c;
        }
        first_guess[WORD_SIZE - 1] = '\0';

        for (int i = 0; i < GUESSES_LENGTH; i++)
        {
            if (strcmp(guesses[i], first_guess) == 0) // if we find the input in guesses return it
                return first_guess;
        }

        printf("** invalid first guess **\n"); // if we haven't found the input in guesses, print error message and retry
    }
}

/**
 * Get the result from the last guess from the user's input.
 * @return Pointer to the result array.
 */
int *get_result()
{
    static int result[WORD_SIZE - 1];
    while (1)
    {
        printf("RESULT: ");
        int error_detected = 0; // to detect when an invalid character is found
        for (int i = 0; i < WORD_SIZE - 1; i++)
        {
            char c = to_lower(getchar());
            if (c == '\n') // skip end line character
            {
                i--;
                continue;
            }

            if (c == GREEN_INPUT)
                result[i] = GREEN;
            else if (c == ORANGE_INPUT)
                result[i] = ORANGE;
            else if (c == BLACK_INPUT)
                result[i] = BLACK;
            else
                error_detected = 1; // invalid character was found
        }
        if (error_detected == 0)
            return result;                // if input was valid, return the result
        printf("** invalid result **\n"); // otherwise print error message and retry
    }
}

/**
 * Get the valid answers and valid guesses from their respective files.
 * @param answers The array to store the valid answers.
 * @param guesses The array to store the valid guesses.
 */
void get_answers_guesses(char answers[][WORD_SIZE], char guesses[][WORD_SIZE])
{
    FILE *f = fopen(ANSWERS_FILE, "r");
    for (int i = 0; i < ANSWER_LENGTH; i++)
    {
        char other[WORD_SIZE]; // use to store empty new line string
        fgets(answers[i], WORD_SIZE, f);
        strcpy(guesses[i], answers[i]);
        fgets(other, WORD_SIZE, f);
    }
    fclose(f);

    f = fopen(GUESSES_FILE, "r");
    for (int i = 0; i < GUESSES_LENGTH; i++)
    {
        char other[WORD_SIZE]; // use to store empty new line string
        fgets(guesses[i + ANSWER_LENGTH], WORD_SIZE, f);
        fgets(other, WORD_SIZE, f);
    }
    fclose(f);
}

/**
 * Get all possible outcomes of any given word.
 * @param outcomes The array to store all possible outcomes.
 */
void get_all_possible_outcomes(int outcomes[TOTAL_OUTCOMES][WORD_SIZE - 1])
{
    for (int i = 0; i < TOTAL_OUTCOMES; i++)
    {
        int num = i;
        for (int j = 0; j < WORD_SIZE - 1; j++) // get the ternary representation of i for outcomes[i]
        {
            outcomes[i][WORD_SIZE - 1 - j - 1] = num % OUTCOMES_LENGTH;
            num /= OUTCOMES_LENGTH;
        }
    }
}

/**
 * Check if an answer is possible, given a guess and its outcome.
 * @param answer The answer to check.
 * @param guess The guess.
 * @param outcome The outcome.
 * @return Boolean indicating whether the answer is possible.
 */
bool is_possible(char answer[WORD_SIZE], char guess[WORD_SIZE], int outcome[WORD_SIZE - 1])
{
    int order[OUTCOMES_LENGTH][WORD_SIZE]; // compartmentalise the outcomes
    int a = 0, b = 0, c = 0;
    for (int i = 0; i < WORD_SIZE - 1; i++)
    {
        if (outcome[i] == GREEN)
            order[0][a++] = i; // green at index 0
        else if (outcome[i] == ORANGE)
            order[1][b++] = i; // orange at index 1
        else
            order[2][c++] = i; // blacks at index 2
    }
    // add a null index to each array to indicate where to stop
    order[0][a] = NULL_INDEX;
    order[1][b] = NULL_INDEX;
    order[2][c] = NULL_INDEX;

    int checked[ALPHABET_SIZE] = {}; // store the characters that have been checked

    for (int i = 0; i < OUTCOMES_LENGTH; i++)
    {
        int j = 0;
        while (order[i][j] != NULL_INDEX)
        {
            int index = order[i][j];
            int count = count_chr(answer, guess[index]);
            int checked_index = guess[index] - 'a' + 1; // index for the checked array

            // if the outcome is green, check if answer contains guess[index] at position index and add to checked
            if (outcome[index] == GREEN)
            {
                if (answer[index] != guess[index]) // answer must contain the letter at the index
                    return false;

                checked[checked_index]++;
            }

            // if outcome is orange, two scenarios are possible
            // - the letter hasn't been checked yet, therefore if the answer contains this letter or has the letter at index it is invalid
            // - the letter has been checked, therefore if we already checked the amount of the letter or the answer contains this letter at index it is invalid
            else if (outcome[index] == ORANGE)
            {
                if (checked[checked_index] == 0) // no greens of this letter have been checked yet for the letter in question
                {
                    if (count == 0 || guess[index] == answer[index]) // answer must contain the letter, not at the same index
                        return false;
                }

                else // some greens of this letter have been checked
                {
                    if (checked[checked_index] == count || guess[index] == answer[index])
                        return false; // answer must contain only count of the letter, and not at the same index
                }
                checked[checked_index]++;
            }

            // if outcome is black, three scenarios are possible
            // - there is no count of the letter in question, which means this answer is valid so far
            // - there is count of the letter and we haven't already checked this letter, meaning this answer is invalid
            // - there is count of the letter and we have checked the letter, and the number of checked is less than the count, meaning answer is invalid
            else
            {
                if (count != 0) // there is count of the letter
                {
                    if (checked[checked_index] == 0) // if no greens or oranges of this letter checked yet
                        return false;
                    else // greens or oranges of this letter have been checked
                    {
                        if (checked[checked_index] < count) // ensure we are not below the count of the letter
                        {
                            return false;
                        }
                    }
                }
            }
            j++;
        }
    }

    return true; // answer passes validation
}

/**
 * Find the next best guess.
 * @param guesses The array of all valid guesses.
 * @param answers The array of all valid answers.
 * @param outcomes The array of possible outcomes.
 * @param answer_indexes The array of indexes of the available answers in the current game state.
 * @return The guess that minimises the average amount of answers left.
 */
char *find_best_guess(char guesses[GUESSES_LENGTH][WORD_SIZE], char answers[ANSWER_LENGTH][WORD_SIZE], int outcomes[TOTAL_OUTCOMES][WORD_SIZE - 1], int answer_indexes[ANSWER_LENGTH + 1])
{
    double min_average = ANSWER_LENGTH; // initialise minimum average
    static char min_word[WORD_SIZE];    // initialise minimum word

    for (int i = 0; i < GUESSES_LENGTH; i++) // for each guess
    {
        int total = TOTAL_OUTCOMES; // initialise total number of valid outcomes
        int possibles = 0;          // initialise number of possible answers

        for (int j = 0; j < TOTAL_OUTCOMES; j++) // for each outcome
        {
            int answers_left = 0; // initialise number of answers left for this outcome
            int k = 0;
            while (answer_indexes[k] != NULL_INDEX) // for each available answer check if it is possible with the current guess and outcome
            {
                if (is_possible(answers[answer_indexes[k]], guesses[i], outcomes[j]))
                    answers_left++; // if so add to number of answers left for this outcome
                k++;
            }

            if (answers_left == 0)     // if we have no answers left for this outcome
                total--;               // valid outcomes decreases
            possibles += answers_left; // add to total number of possible answers
        }

        double average_words_left = (double)possibles / total; // get the average number of answers left for this guess
        if (average_words_left == 1)
            return guesses[i]; // if average is 1, shortcut and return immediately (can't get better)

        if (average_words_left < min_average) // otherwise, update min_average, and min_word if out new average is smaller than min_average
        {
            strcpy(min_word, guesses[i]);
            min_average = average_words_left;
        }

        printf("Checking %d/%d\r", i + 1, GUESSES_LENGTH); // print current state of processing
        fflush(stdout);                                    // flush stdout for cleaner '\r' printing
    }

    return min_word; // return the best word found
}

/**
 * Filter the currently available answers to account for a new guess and result.
 * @param answers The array of all possible answers.
 * @param answer_indexes The indexes of the currently available answers for the current game state.
 * @param result The result of the new guess.
 * @param guess The new guess.
 * @return The total amount of viable answers left.
 */
int filter_answers(char answers[ANSWER_LENGTH][WORD_SIZE], int answer_indexes[ANSWER_LENGTH + 1], int result[WORD_SIZE - 1], char current_guess[WORD_SIZE])
{
    int new_indexes[ANSWER_LENGTH + 1]; // new indexes array
    int new_length = 0;
    int i = 0;

    while (answer_indexes[i] != NULL_INDEX) // loop through the currently available answers
    {
        int index = answer_indexes[i];
        if (is_possible(answers[index], current_guess, result)) // if the answer is still possible add to new indexes array
            new_indexes[new_length++] = index;
        i++;
    }
    new_indexes[new_length] = NULL_INDEX;

    int j = 0;
    while (new_indexes[j] != NULL_INDEX) // copy the new indexes array back into answer indexes
    {
        answer_indexes[j] = new_indexes[j];
        j++;
    }
    answer_indexes[j] = NULL_INDEX;

    return new_length; // return the total number of answers left
}

/**
 * Solve for a word.
 * @param total_guesses The total number of guesses so far.
 * @param current_guess The current guess for the word.
 * @param answers The array of all possible answers.
 * @param guesses The array of all possible guesses.
 * @param outcomes The array of all possible outcomes.
 * @param answer_indexes The array of indexes of all currently available answers in the game state.
 * @return The solution, or NULL if an error occurred.
 */
char *solve_word(int *total_guesses, char current_guess[WORD_SIZE], char answers[ANSWER_LENGTH][WORD_SIZE], char guesses[GUESSES_LENGTH][WORD_SIZE], int all_outcomes[TOTAL_OUTCOMES][WORD_SIZE - 1], int answer_indexes[ANSWER_LENGTH + 1])
{
    (*total_guesses)++; // increment the total guesses
    int *result = get_result(); // get the result from the current guess

    int answers_left = filter_answers(answers, answer_indexes, result, current_guess); // filter the answers to find how many answers are left
    if (result[0] == GREEN && result[1] == GREEN && result[2] == GREEN && result[3] == GREEN && result[4] == GUESSES_LENGTH)
    { // if the current guess is correct, return it and decrement total guesses (as we found it before the next guess)
        (*total_guesses)--;
        return current_guess;
    }
    else // otherwise print the amount of answers left
        printf("-- answers left: %d --\n\n", answers_left);

    if (answers_left == 0) // if we have no answers left, an error has occured so return null
        return NULL;
    else if (answers_left == 1) // if we have one answer left, return it
        return answers[answer_indexes[0]];

    char *next_best = find_best_guess(guesses, answers, all_outcomes, answer_indexes); // otherwise find the next best guess
    printf(" GUESS: %s         \n", next_best); // print it out
    return solve_word(total_guesses, next_best, answers, guesses, all_outcomes, answer_indexes); // keep solving
}

int main()
{
    char answers[ANSWER_LENGTH][WORD_SIZE];
    char guesses[GUESSES_LENGTH][WORD_SIZE];
    get_answers_guesses(answers, guesses); // get answers and guesses

    char *first_guess = get_first_guess(guesses); // get the first guess

    int all_outcomes[TOTAL_OUTCOMES][WORD_SIZE - 1];
    get_all_possible_outcomes(all_outcomes); // get all outcomes

    int answer_indexes[ANSWER_LENGTH + 1]; // initialise answer indexes that stores the indexes of all currently available answers
    for (int i = 0; i < ANSWER_LENGTH; i++)
        answer_indexes[i] = i;
    answer_indexes[ANSWER_LENGTH] = NULL_INDEX;

    int total_guesses = 1; // we've had one guess already
    char *solution = solve_word(&total_guesses, first_guess, answers, guesses, all_outcomes, answer_indexes); // solve for the word

    if (solution == NULL) // if invalid results were inputted that lead to no answers left
    {
        printf("** ERROR: no answers left - invalid result(s) **\n");
        return 1;
    }
    printf("** ANSWER: %s - Guesses: %d **\n\n", solution, total_guesses); // print the summary of the solution
    return 0;
}
